File: web/server/store.js
"use strict";

const fs = require("fs");
const path = require("path");

const C = require("./crypto");

function atomicWriteJson(filePath, obj) {
  const tmp = filePath + ".tmp";
  fs.writeFileSync(tmp, JSON.stringify(obj), "utf8");
  fs.renameSync(tmp, filePath);
}

function readJson(filePath, fallback) {
  try {
    return JSON.parse(fs.readFileSync(filePath, "utf8"));
  } catch (e) {
    return fallback;
  }
}

class SecretBox {
  constructor(dataDir) {
    this.filePath = path.join(dataDir, "secret_key.json");
    let doc = readJson(this.filePath, null);
    if (doc && doc.k) {
      try {
        this.key = C.fromB64(doc.k);
      } catch (e) {
        this.key = null;
      }
    } else {
      this.key = null;
    }
    if (!this.key || this.key.length !== C.KEY_LEN) {
      this.key = C.randomBytes(C.KEY_LEN);
      atomicWriteJson(this.filePath, { scheme: "plain", k: C.toB64(this.key) });
    }
  }

  protect(text) {
    if (!text) return text;
    try {
      return "enc1:" + C.toB64(C.aesGcmEncrypt(this.key, Buffer.from(String(text), "utf8")));
    } catch (e) {
      return text;
    }
  }

  unprotect(text) {
    if (!text || !String(text).startsWith("enc1:")) return text;
    try {
      return C.aesGcmDecrypt(this.key, C.fromB64(String(text).slice(5))).toString("utf8");
    } catch (e) {
      return "";
    }
  }
}

class Store {
  constructor(dataDir) {
    this.dataDir = dataDir;
    this.chatsDir = path.join(dataDir, "chats");
    fs.mkdirSync(this.chatsDir, { recursive: true });
    this.secretBox = new SecretBox(dataDir);
  }

  _path(name) {
    return path.join(this.dataDir, name);
  }

  load(name, fallback) {
    return readJson(this._path(name), fallback);
  }

  save(name, obj) {
    atomicWriteJson(this._path(name), obj);
  }

  chatPath(key) {
    const safe = Buffer.from(key, "utf8").toString("hex");
    return path.join(this.chatsDir, safe + ".json");
  }

  loadChat(key) {
    const doc = readJson(this.chatPath(key), null);
    if (!doc || !Array.isArray(doc.messages)) return [];
    return doc.messages;
  }

  saveChat(key, messages) {
    const msgs = messages.map((m) => {
      const out = Object.assign({}, m);
      if (out.content != null && !(typeof out.content === "string" && out.content.startsWith("enc1:"))) {
        out.content = this.secretBox.protect(out.content);
      }
      if (out.replyPreview) out.replyPreview = this.secretBox.protect(out.replyPreview);
      return out;
    });
    atomicWriteJson(this.chatPath(key), { messages: msgs });
  }

  loadChatDecrypted(key) {
    return this.loadChat(key).map((m) => {
      const out = Object.assign({}, m);
      if (typeof out.content === "string" && out.content.startsWith("enc1:")) {
        out.content = this.secretBox.unprotect(out.content);
      }
      if (typeof out.replyPreview === "string" && out.replyPreview.startsWith("enc1:")) {
        out.replyPreview = this.secretBox.unprotect(out.replyPreview);
      }
      return out;
    });
  }

  deleteChat(key) {
    try {
      fs.unlinkSync(this.chatPath(key));
    } catch (e) {
      /* ignore */
    }
  }

  protect(text) {
    return this.secretBox.protect(text);
  }

  unprotect(text) {
    return this.secretBox.unprotect(text);
  }
}

module.exports = { Store, SecretBox, atomicWriteJson, readJson };

File: server/webchat/store.py
import json
import os

from . import crypto as C


def atomic_write_json(file_path, obj):
    tmp = file_path + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        f.write(json.dumps(obj, separators=(",", ":"), ensure_ascii=False))
    os.replace(tmp, file_path)


def read_json(file_path, fallback):
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            return json.loads(f.read())
    except (OSError, ValueError):
        return fallback


class SecretBox:
    def __init__(self, data_dir):
        self.file_path = os.path.join(data_dir, "secret_key.json")
        doc = read_json(self.file_path, None)
        self.key = None
        if isinstance(doc, dict) and doc.get("k"):
            try:
                self.key = C.from_b64(doc["k"])
            except Exception:
                self.key = None
        if not isinstance(self.key, (bytes, bytearray)) or len(self.key) != C.KEY_LEN:
            self.key = C.random_bytes(C.KEY_LEN)
            atomic_write_json(self.file_path, {"scheme": "plain", "k": C.to_b64(self.key)})
        self.key = bytes(self.key)

    def protect(self, text):
        if not text:
            return text
        try:
            return "enc1:" + C.to_b64(C.aes_gcm_encrypt(self.key, str(text).encode("utf-8")))
        except Exception:
            return text

    def unprotect(self, text):
        if not text or not str(text).startswith("enc1:"):
            return text
        try:
            return C.aes_gcm_decrypt(self.key, C.from_b64(str(text)[5:])).decode("utf-8")
        except Exception:
            return ""


class Store:
    def __init__(self, data_dir):
        self.data_dir = data_dir
        self.chats_dir = os.path.join(data_dir, "chats")
        os.makedirs(self.chats_dir, exist_ok=True)
        self.secret_box = SecretBox(data_dir)

    def _path(self, name):
        return os.path.join(self.data_dir, name)

    def load(self, name, fallback):
        return read_json(self._path(name), fallback)

    def save(self, name, obj):
        atomic_write_json(self._path(name), obj)

    def chat_path(self, key):
        safe = key.encode("utf-8").hex()
        return os.path.join(self.chats_dir, safe + ".json")

    def load_chat(self, key):
        doc = read_json(self.chat_path(key), None)
        if not isinstance(doc, dict) or not isinstance(doc.get("messages"), list):
            return []
        return doc["messages"]

    def save_chat(self, key, messages):
        msgs = []
        for m in messages:
            out = dict(m)
            content = out.get("content")
            if content is not None and not (
                isinstance(content, str) and content.startswith("enc1:")
            ):
                out["content"] = self.secret_box.protect(content)
            if out.get("replyPreview"):
                out["replyPreview"] = self.secret_box.protect(out["replyPreview"])
            msgs.append(out)
        atomic_write_json(self.chat_path(key), {"messages": msgs})

    def load_chat_decrypted(self, key):
        out = []
        for m in self.load_chat(key):
            item = dict(m)
            content = item.get("content")
            if isinstance(content, str) and content.startswith("enc1:"):
                item["content"] = self.secret_box.unprotect(content)
            preview = item.get("replyPreview")
            if isinstance(preview, str) and preview.startswith("enc1:"):
                item["replyPreview"] = self.secret_box.unprotect(preview)
            out.append(item)
        return out

    def delete_chat(self, key):
        try:
            os.unlink(self.chat_path(key))
        except OSError:
            pass

    def protect(self, text):
        return self.secret_box.protect(text)

    def unprotect(self, text):
        return self.secret_box.unprotect(text)
